information/specification.md

```diff
@@ -0,0 +1,826 @@
+    list](#requested-output-list), sorted into _descending order of coin
+    value_.
+
+    The _head_ of the list is always the unpaid output with the _largest coin
+    value_.
+
+    Entries are incrementally removed from the _head_ of the list as the
+    algorithm proceeds, until the list is empty.
+
+ 3. #### Accumulated Coin Selection
+
+    The accumulated coin selection is a [coin selection](#coin-selection) where
+    all fields are initially equal to the _empty set_.
+
+### Computation
+
+The algorithm proceeds according to the following sequence of steps:
+
+ * **Step 1**
+
+   Remove a single _unpaid output_ from the head of the [unpaid output
+   list](#unpaid-output-list).
+
+ * **Step 2**
+
+   Repeatedly remove UTxO entries from the head of the [available UTxO
+   list](#available-utxo-list) until the total value of entries removed is
+   _greater than or equal to_ the value of the _removed output_.
+
+ * **Step 3**
+
+   Use the _removed UTxO entries_ to pay for the _removed output_.
+
+   This is achieved by:
+
+    * adding the _removed UTxO entries_ to the _inputs_ field of the
+      [accumulated coin selection](#accumulated-coin-selection).
+    * adding the _removed output_ to the _outputs_ field of the
+      [accumulated coin selection](#accumulated-coin-selection).
+
+ * **Step 4**
+
+   If the _total value_ of the _removed UTxO entries_ is greater than the
+   value of the _removed output_, generate a coin whose value is equal to
+   the exact difference, and add it to the _change values_ field of the
+   [accumulated coin selection](#accumulated-coin-selection).
+
+ * **Step 5**
+
+   If the _unpaid output list_ is empty, **terminate** here.
+
+   Otherwise, return to **Step 1**.
+
+### Termination
+
+The algorithm terminates _successfully_ if the [available UTxO
+list](#available-utxo-list) is not depleted before the [unpaid output
+list](#unpaid-output-list) can be fully depleted (i.e., if all the outputs have
+been paid for).
+
+The [accumulated coin selection](#accumulated-coin-selection) is returned
+to the caller as the [coin selection](#coin-selection) result.
+
+The [available UTxO list](#available-utxo-list) is returned to the caller
+as the [remaining UTxO set](#remaining-utxo-set) result.
+
+## Random-Improve
+
+The **Random-Improve** coin selection algorithm works in _two phases_.
+
+In the _first_ phase, the algorithm iterates through each of the [requested
+outputs](#requested-output-list) in _descending order of coin value_, from
+_largest_ to _smallest_. For each output, the algorithm repeatedly selects
+entries at **random** from the [initial UTxO set](#initial-utxo-set), until the
+_total value_ of selected entries is enough to pay for that ouput.
+
+In the _second_ phase, the algorithm attempts to **improve** upon each of the
+UTxO selections made in the previous phase, by conservatively expanding the
+selection made for each output, in order to generate improved change values.
```

Comment:
It'll be worth mentioning here what a "improved" output is and why is that so:

- A new change value is considered an improvement if its value gets _closer_ to the value of the corresponding output being covered in the first place. 

The algorithm works by considering, for each output, the triple (minValue, idealValue, maxValue) of values to cover. `minValue` is obviously the value of the selected output itself and, to be considered successful, the algorithm **must* select inputs to cover for at least this `minValue`. And then, improve to try covering for `idealValue`, but no more than `maxValue` (otherwise, we would end up selecting a lot more than what's really needed). In practice, `idealValue` is set to twice the output value and `maxValue` to thrice the output value. 
This therefore aims for constructing change addresses that are roughly equal to the output value (with an error margin or +/- 1 output value; which turns out to be good enough in simulations and empirical observations). 
On large UTxO sets where the variety of coins is bigger, one could imagine using more restrictive constraints like:

- min = 1.75 * output
- ideal = 2*output 
- max = 2.25 * output

such that change outputs are roughly of the same size as the output but with a greater precision.

Reply:
@KtorZ I've extended the description of the algorithm in 4cf7b1334a40d382f2642ca1187a9f12a339f8d6.

The above change attempts to explain the motivation behind attempting to produce a selection for each output that's around twice the value of that output.

Let me know if you think it can be improved further. (I'll also make changes to subsequent sections, where appropriate.)